[iOS] AudioContext is getting suspended when page goes in the background even if navigator.audioSession.type is set to playback

youennf opened the pull request:
#### b848143a2ad57c80405dc5b30e61c750f5896d72
<pre>
[iOS] AudioContext is getting suspended when page goes in the background even if navigator.audioSession.type is set to playback
<a href="https://bugs.webkit.org/show_bug.cgi?id=261554">https://bugs.webkit.org/show_bug.cgi?id=261554</a>
<a href="https://rdar.apple.com/115485355">rdar://115485355</a>

Reviewed by Eric Carlson.

Implement isNowPlayingEligible, nowPlayingInfo and selectBestMediaSession to enable NowPlayingInfo handling.
Only activated by JS AudioContext(s) are NowPlayingInfo-able if DOM AudioSession type is playback or plabyack and record.

Implement canReceiveRemoteControlCommands and didReceiveRemoteControlCommand to be able to play/pause according commands.
Update AudioContext::mayResumePlayback to only resume for activated by JS AudioContext(s).
Update AudioContext::shouldOverrideBackgroundPlaybackRestriction to be able to continue in the background in case of
playback or plabyack and record DOM AudioSession(s).

Update bug in PlatformMediaSessionManager::bestEligibleSessionForRemoteControls late time optimization to only add web audio sessions to its list
when there is no media element eligible session. If there is an eligible media element session, we do not want to add web audio session to the media session list.

* LayoutTests/media/now-playing-webaudio-expected.txt: Added.
* LayoutTests/media/now-playing-webaudio.html: Added.
* LayoutTests/media/webaudio-background-playback-expected.txt:
* LayoutTests/media/webaudio-background-playback.html:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/Modules/mediasession/NavigatorMediaSession.cpp:
(WebCore::NavigatorMediaSession::mediaSessionIfExists):
* Source/WebCore/Modules/mediasession/NavigatorMediaSession.h:
* Source/WebCore/Modules/webaudio/AudioContext.cpp:
(WebCore::AudioContext::AudioContext):
(WebCore::AudioContext::mayResumePlayback):
(WebCore::AudioContext::canReceiveRemoteControlCommands const):
(WebCore::AudioContext::didReceiveRemoteControlCommand):
(WebCore::hasPlayBackAudioSession):
(WebCore::AudioContext::isNowPlayingEligible const):
(WebCore::AudioContext::nowPlayingInfo const):
(WebCore::AudioContext::selectBestMediaSession):
(WebCore::AudioContext::shouldOverrideBackgroundPlaybackRestriction const):
(WebCore::AudioContext::defaultDestinationWillBecomeConnected):
* Source/WebCore/Modules/webaudio/AudioContext.h:
* Source/WebCore/Modules/webaudio/BaseAudioContext.cpp:
(WebCore::BaseAudioContext::setState):
* Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp:
(WebCore::PlatformMediaSessionManager::bestEligibleSessionForRemoteControls):

Canonical link: <a href="https://commits.webkit.org/275558@main">https://commits.webkit.org/275558@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9e25fca4ac0bb1c74c6ec605bddfafd97d583da5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [⏳ 🧪 style ](https://ews-build.webkit.org/#/builders/Style-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-17-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wincairo ](https://ews-build.webkit.org/#/builders/WinCairo-EWS "Waiting in queue, processing has not started yet") 
| [⏳ 🧪 bindings ](https://ews-build.webkit.org/#/builders/Bindings-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Sonoma-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 webkitperl ](https://ews-build.webkit.org/#/builders/WebKitPerl-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-mac ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 webkitpy ](https://ews-build.webkit.org/#/builders/WebKitPy-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2-wpt ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk1 ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc ](https://ews-build.webkit.org/#/builders/JSC-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 gtk-wk2 ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-arm64-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-17-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Sonoma-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 services ](https://ews-build.webkit.org/#/builders/Services-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2-stress ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9443 "Built successfully and passed tests") | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-10-Build-EWS "Waiting in queue, processing has not started yet") | | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-10-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson reviewed (approved):
I think this is good, but the commit log should have a longer explanation of _why_ the change is being made.

jyavenard commented on LayoutTests/media/now-playing-webaudio.html:
```diff
@@ -0,0 +1,104 @@
+    </head>
+    <body>
+        <script>
+function waitFor(delay)
```
that's in video-test.js already

youennf replied:
video-test.js is not test harness based, so I prefer not mixing things.

jyavenard commented on LayoutTests/media/now-playing-webaudio.html:
```diff
@@ -0,0 +1,104 @@
+    return new Promise(resolve => setTimeout(resolve, delay));
+}
+
+async function waitForCriteria(test, criteria)
```
this one too, there's `testExpectedEventually`

youennf replied:
Ditto

jyavenard commented on LayoutTests/media/now-playing-webaudio.html:
```diff
@@ -0,0 +1,104 @@
+
+    let context = new AudioContext();
+
+    waitFor(100);
```
no await ? otherwise this serves no purpose

youennf replied:
Right, will fix (or remove).

youennf replied:
Fixed

jyavenard commented on LayoutTests/media/now-playing-webaudio.html:
```diff
@@ -0,0 +1,104 @@
+
+    context.resume();
+
+    await waitForCriteria(test, () => {
```
await `testExpectedEventually`

jyavenard commented on Source/WebCore/Modules/webaudio/AudioContext.cpp:
```diff
@@ -507,14 +643,17 @@ bool AudioContext::shouldOverrideBackgroundPlaybackRestriction(PlatformMediaSess
 
-    return !destination().isConnected();
+    RefPtr document = this->document();
+    return hasPlayBackAudioSession(RefPtr { this->document() }.get());
```
that time convoluted; what's `document` for here ?

youennf replied:
Oh, yes it should be hasPlayBackAudioSession(document.get()).

youennf replied:
Fixed

eric-carlson commented on LayoutTests/platform/glib/TestExpectations:
```diff
@@ -1,4 +1,4 @@
-# These are the layout test expectations for the glib-based ports: WebKitGTK and WPE.
+await # These are the layout test expectations for the glib-based ports: WebKitGTK and WPE.
```
Is this intentional?

youennf replied:
nope, thanks!

youennf replied:
fixed

youennf committed:
[iOS] AudioContext is getting suspended when page goes in the background even if navigator.audioSession.type is set to playback
https://bugs.webkit.org/show_bug.cgi?id=261554
rdar://115485355

Reviewed by Eric Carlson.

Implement isNowPlayingEligible, nowPlayingInfo and selectBestMediaSession to enable NowPlayingInfo handling.
Only activated by JS AudioContext(s) are NowPlayingInfo-able if DOM AudioSession type is playback or plabyack and record.

Implement canReceiveRemoteControlCommands and didReceiveRemoteControlCommand to be able to play/pause according commands.
Update AudioContext::mayResumePlayback to only resume for activated by JS AudioContext(s).
Update AudioContext::shouldOverrideBackgroundPlaybackRestriction to be able to continue in the background in case of
playback or plabyack and record DOM AudioSession(s).

Update bug in PlatformMediaSessionManager::bestEligibleSessionForRemoteControls late time optimization to only add web audio sessions to its list
when there is no media element eligible session. If there is an eligible media element session, we do not want to add web audio session to the media session list.

* LayoutTests/media/now-playing-webaudio-expected.txt: Added.
* LayoutTests/media/now-playing-webaudio.html: Added.
* LayoutTests/media/webaudio-background-playback-expected.txt:
* LayoutTests/media/webaudio-background-playback.html:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/Modules/mediasession/NavigatorMediaSession.cpp:
(WebCore::NavigatorMediaSession::mediaSessionIfExists):
* Source/WebCore/Modules/mediasession/NavigatorMediaSession.h:
* Source/WebCore/Modules/webaudio/AudioContext.cpp:
(WebCore::AudioContext::AudioContext):
(WebCore::AudioContext::mayResumePlayback):
(WebCore::AudioContext::canReceiveRemoteControlCommands const):
(WebCore::AudioContext::didReceiveRemoteControlCommand):
(WebCore::hasPlayBackAudioSession):
(WebCore::AudioContext::isNowPlayingEligible const):
(WebCore::AudioContext::nowPlayingInfo const):
(WebCore::AudioContext::selectBestMediaSession):
(WebCore::AudioContext::shouldOverrideBackgroundPlaybackRestriction const):
(WebCore::AudioContext::defaultDestinationWillBecomeConnected):
* Source/WebCore/Modules/webaudio/AudioContext.h:
* Source/WebCore/Modules/webaudio/BaseAudioContext.cpp:
(WebCore::BaseAudioContext::setState):
* Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp:
(WebCore::PlatformMediaSessionManager::bestEligibleSessionForRemoteControls):

Canonical link: https://commits.webkit.org/275558@main
